### PR TITLE
Add react-is package

### LINKED ISSUE #12038: Add React.isFragment api for verifying Fragment
**Do you want to request a *feature* or report a *bug*?**

I want to request a feature.

**What is the current behavior?**

We have no API to verify a ReactNode is a React 16 Fragment. Though now we can use `React.isValidElement(instance) && typeof instance.type === 'symbol'` to distinguish it. It's verbose and seems uncertianly right.

**What is the expected behavior?**

Add api:

```
React.isFragment(object)
```
Verifies the object is a React.Fragment. Returns true or false.

**Which versions of React, and which browser / OS are affected by this issue? Did this work in previous versions of React?**

React 16.2

### PR DESCRIPTION
**Pull request by bvaughn**
Supersedes PR #11279 and #12092. Resolves issue #12038

~~Perhaps we could use this in React DevTools as well?~~ (This would not be the case in its current form, since this library accepts React _elements_ and DevTools receives _fibers_.)

### Tests

In addition to the newly-added unit tests, I ran `yarn build react-is` locally to build this, then `npm pack`ed the result and installed it in a fresh project. I also tested the UMD build in CodeSandbox.

### Example
```js
import React from "react";
import {
  ContextConsumer,
  ContextProvider,
  isContextConsumer,
  isContextProvider,
  typeOf
} from "react-is";

const ThemeContext = React.createContext("blue");

isContextConsumer(<ThemeContext.Consumer />); // true
isContextProvider(<ThemeContext.Provider />); // true
typeOf(<ThemeContext.Provider />) === ContextProvider; // true
typeOf(<ThemeContext.Consumer />) === ContextConsumer; // true
```

### Assumptions
* It is okay that `isElement` is looser than `typeOf` (e.g. it returns true for host elements, fragments, strict-mode, context, etc.)
* There are no additional component types (e.g. the experimental call/return types) that we want to expose for the initial release.
* It's okay to support the (currently unstable) `AsyncMode` for the initial release.
* This package should mirror the version of React rather than be versioned independently (since we'll be adding types over time).

### CODE REVIEW

**Comment by sebmarkbage** [comment]
I'm personally not a big fan of the `typeOf` mechanism, but not strongly enough to block it.

Another plausible approach is that we expose the symbols and `typeOf` returns a symbol. You can get the string name through `String(typeOf(...))` and in the future, the `symbol.description` proposal.

Then you can switch on symbols instead of strings:

```js
switch (typeOf(node)) {
  case ReactIs.Element:
    ...
  case ReactIs.Fragment:
    ...
}
```

**Comment by bvaughn** [comment]
> Another plausible approach is that we expose the symbols and `typeOf` returns a symbol. You can get the string name through `String(typeOf(...))` and in the future, the `symbol.description` proposal.

Wouldn't the issue still remain that we'd have to access the symbols from a different attribute/path depending on the type of element (e.g. `object.type.$$typeof` for context vs `object.type` for fragments vs `object.$$typeof` for portals)

**Comment by sebmarkbage** [comment]
The implementation of typeOf would work exactly the same. It would just return symbols as the return value instead of strings.

I don’t know if I can really motivate this very strongly though. Just feels more specific with symbols since they don’t take up string space but in practice strings are probably fine and more practical.

**Comment by bvaughn** [comment]
Oh, I see. I misunderstood your initial suggestion 😄

So you're saying rather than someone using e.g.
```js
import React from 'react';
import ReactIs from 'react-is';

const ThemeContext = React.createContext('blue');

ReactIs.isContextConsumer(<ThemeContext.Consumer />);
ReactIs.isContextProvider(<ThemeContext.Provider />);
```

They would use:
```js
import React from 'react';
import ReactIs from 'react-is';

const ThemeContext = React.createContext('blue');

ReactIs.typeOf(<ThemeContext.Provider />) === ReactIs.ContextProvider;
ReactIs.typeOf(<ThemeContext.Consumer />) === ReactIs.ContextConsumer;
```

I guess we'd still have to export the numeric fallback values for environments that don't support `Symbol.for` (like IE and RN Android)?

I'm not opposed to making this small change, if that's what's required for this lib to be shippable. 😄

**Comment by sebmarkbage** [comment]
I think using isContextConsumer is still best practice when you only test for a single value.

The typeOf is more about logging and switch statements where there are many possibly legit options. We stopped using isValidElement internally for the benefit of switch statements so seems like others might want to too. Although strings support that case too.

**Comment by bvaughn** [comment]
Cool. Seems like exporting constants for the string return-types would satisfy the switch/case use case as well.

I don't really have a strong feeling. Gimme a few and I'll push an update.

**Comment by bvaughn** [comment]
Okedoke. Rebased and added symbol exports.

**Comment by bvaughn** [comment]
One thing I'm not clear about is the `isElement`.

I think your original PR would have returned true for more things that my current PR does. Is this okay or should I change my implementation so `isElement` returns true for e.g. fragments and context as well?

**Comment by bvaughn** [comment]
Actually, disregard my previous comment. 😄

File: packages/react-is/index.js [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+ * @flow
+ */
+
+'use strict';
+
+const ReactIs = require('./src/ReactIs');
+
+// TODO: decide on the top-level export form.
+// This is hacky but makes it work with both Rollup and Jest.
+module.exports = ReactIs.default ? ReactIs.default : ReactIs;
```
Comment (gaearon) [inline packages/react-is/index.js]: We should do this correctly right away since it's not a legacy package.

Let's remove the default export here and only provide named exports. That's how we make it work with tree shaking, too (when we provide an ES build output).

Comment (gaearon) [reply packages/react-is/index.js]: To clarify, I'm saying this file shouldn't need to use CommonJS at all. Just re-export `*` from the source.

Comment (bvaughn) [reply packages/react-is/index.js]: Hm. Okay.

Why didn't we follow that pattern with other, brand-new packages (like `react-reconciler`)?

Comment (gaearon) [reply packages/react-is/index.js]: I think there was some reason but I don't remember straight away.

It's not very important in case of the reconciler though because it provides just one function. `react-is`, on the other hand, provides a bunch, and is expected to grow.

Call/Return does avoid the `.default` hack: https://github.com/facebook/react/blob/29e8924c70856bef9b11e0c74a450140bfcce773/packages/react-call-return/index.js#L12

If there's some issue with top-level ES import (maybe there was) then the above is sufficient.

Comment (bvaughn) [reply packages/react-is/index.js]: Okay. This makes sense. Thanks for the suggestion. 😄

It's been updated.

**Comment by bvaughn** [comment]
It looks like the Rollup build checks report an ESLint failure from the UMD builds:
> /home/circleci/project/build/node_modules/react-is/umd/react-is.development.js
  13:73  error  'exports' is not defined  no-undef

The line in question is:
```js
typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports) :
```

This is the only UMD bundle that uses the `exports` object as a param to the function.

I believe we could fix the lint error (without breaking the UMD build) by changing the signature to:
```js
typeof module === 'object' && typeof module.exports === 'object' ? factory(module.exports) :
```

But I'm not sure how to achieve this with our current build scripts. 😄 I'll poke around some.

I'm also not convinced a UMD build is strictly necessary for this module. It just seemed nice to have.

File: packages/react-is/README.md [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+isElement(<div />); // true
+typeOf(<div />) === Element; // true
+```
+
+### Fragment
+
+```js
+import React from "react";
+import { Fragment, isFragment, typeOf } from "react-is";
```
Comment (gaearon) [inline packages/react-is/README.md]: I could see this being confusing (e.g. when you read `Fragment` you assume it's from `react`).

Maybe let's keep it grouped, like before? It just needs to be a namespace import rather than a default one.

```js
import * as ReactIs from 'react-is';
```

**Comment by gaearon** [comment]
>It looks like the Rollup build checks report an ESLint failure from the UMD builds:

We can add it [here](https://github.com/facebook/react/blob/29e8924c70856bef9b11e0c74a450140bfcce773/scripts/rollup/validate/eslintrc.umd.js#L19). The list is not exhaustive, it just didn't happen to use `exports` in other builds so I didn't whitelist it.

**Comment by bvaughn** [comment]
Okay.

File: packages/react-is/src/ReactIs.js [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    : null;
+}
+
+export function typeOf(object: any) {
+  let maybeType = getType(object);
+  switch (maybeType) {
+    case REACT_ASYNC_MODE_TYPE:
+    case REACT_FRAGMENT_TYPE:
+    case REACT_STRICT_MODE_TYPE:
+      return maybeType;
```
Comment (gaearon) [inline packages/react-is/src/ReactIs.js]: This seems like it would return true for `{type: 0xeacb}`. But React wouldn't consider this a fragment. It would only consider `{type: 0xeacb, $$typeof: 0xeac7}` a fragment. Same for all other things with `type`. This is a small detail but IMO we should be strict about these things.

Comment (bvaughn) [reply packages/react-is/src/ReactIs.js]: Yeah, that's reasonable.

File: packages/react-is/src/ReactIs.js [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+  }
+
+  maybeType = getTypeOf(object);
+  switch (maybeType) {
+    case REACT_ELEMENT_TYPE:
+    case REACT_PORTAL_TYPE:
+      return maybeType;
+  }
+}
```
Comment (gaearon) [inline packages/react-is/src/ReactIs.js]: I find the structure of this function a bit hard to follow. Is there any specific reason it is written this way?
It also seems like we're reading `.type` and compare it many times in pretty common cases like passing an element.

I would prefer to write it inline:

```js
export function typeOf(object: any) {
  const $$typeof = getTypeOf(object);
  switch ($$typeof) {
    case REACT_ELEMENT_TYPE:
      const type = object.type;
      switch (type) {
        case REACT_ASYNC_MODE_TYPE:
        case REACT_FRAGMENT_TYPE:
        case REACT_STRICT_MODE_TYPE:
          return type;
        default:
          const $$typeofType = type.$$typeof;
          switch ($$typeofType) {
            case REACT_CONTEXT_TYPE:
            case REACT_PROVIDER_TYPE:
              return $$typeofType;
            default:
              return $$typeof;
          }
      }
    case REACT_PORTAL_TYPE:
      return $$typeof;
    default:
      return null;
  }
}
```

What do you think? IMO this makes the object structure a bit clearer, and avoids unnecessary property reads.

File: packages/react-is/src/ReactIs.js [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+export const AsyncMode = REACT_ASYNC_MODE_TYPE;
+export const ContextConsumer = REACT_CONTEXT_TYPE;
+export const ContextProvider = REACT_PROVIDER_TYPE;
+export const Element = REACT_ELEMENT_TYPE;
+export const Fragment = REACT_FRAGMENT_TYPE;
+export const Portal = REACT_PORTAL_TYPE;
+export const StrictMode = REACT_STRICT_MODE_TYPE;
+
+export function isAsyncMode(object: any) {
+  return (
```
Comment (gaearon) [inline packages/react-is/src/ReactIs.js]: Seems like all of these could be rewritten as `getTypeOf` comparisons to ensure 1:1 check parity?

Comment (bvaughn) [reply packages/react-is/src/ReactIs.js]: There is a lot of back and forth on a PR, and the latest round seems entirely like nits. Seems a bit inefficient.

WRT using typeOf here- Either way should be fine. It's just a subjective preference, right?

Comment (bvaughn) [reply packages/react-is/src/ReactIs.js]: PS using typeOf in isElement didn't work, because the return type was overly specific. I didn't feel strongly about it, but I opted out of using typeOf director for that reason.

Comment (gaearon) [reply packages/react-is/src/ReactIs.js]: I don't agree that repeating the same checks many times is a nit. Generally we avoid this everywhere else in the codebase. (If you're referring to the previous comment.)

Yes, they probably don't make a difference in a large scheme of things, but if we can make it more efficient, why not do it?

For this comment, I'm not sure. The issue I noted in https://github.com/facebook/react/pull/12199#discussion_r167444125i was a correctness issue. It was fixed in https://github.com/facebook/react/pull/12199/commits/b22a7b78836d7d8252989c7a9a7522c40e7d2637. The fix required changes in two places. My concern is that it's confusing to remember both those places need to contain this logic. When we add a new type it won't be obvious which ones of those `get*` functions need to be called for `is*` to work consistently. Expressing this via `typeof()` solves this problem by consolidating that logic in one place.

Comment (gaearon) [reply packages/react-is/src/ReactIs.js]: >PS using typeOf in isElement didn't work, because the return type was overly specific. I didn't feel strongly about it, but I opted out of using typeOf director for that reason.

You probably considered it but another way to express this would be

```js
switch (typeof(obj)) {
  case Element:
  case ContextConsumer:
  case ContextProvider:
    return true;
  default:
    return false;
}
```

I agree there's no clear winner in this particular case though.

**Comment by gaearon** [comment]
What should `typeof()` return for unexpected inputs? It currently seems to return undefined.
Should it throw? Why (not)?

**Comment by gaearon** [comment]
To clarify what I mean about extra checks and property access.

`typeOf(<div />)` with the current implementation:

* compare `typeof object === 'object'`
* compare `object !== null`
* read `object.type`
* switch on `maybeType`
* compare `typeof object === 'object'`
* compare `object !== null`
* compare `typeof object.type === 'object'`
* compare `object.type !== null`
* read `object.type.$$typeof`
* switch on `maybeType`
* compare `typeof object === 'object'`
* compare `object !== null`
* read `object.$$typeof`
* switch on `maybeType`

`typeOf(<div />)` with implementation in https://github.com/facebook/react/pull/12199#discussion_r167446453:

* compare `typeof object === 'object'`
* compare `object !== null`
* read `object.$$typeof`
* switch on `$$typeof`
* read `object.type`
* read `type.$$typeof`
* switch on `$$typeofType`

We know that `object.type` can't be `null` or `undefined` if its `$$typeof` is an element because we forbid creating elements with null/undefined types. The `object !== null` and `typeof object` is only done once. The `object.type` read isn't repeated.

Ultimately this probably doesn't matter in practice because in most cases the engine will be smart enough to know these couldn't have changed. It probably won't be the hottest path either. Still I hope this helps explain why I felt the existing solution was a bit inelegant. Again, it's not a blocker, just an observation. Thanks for pushing this through!

File: packages/react-is/src/ReactIs.js (hunk 5 above)
Comment (bvaughn) [reply packages/react-is/src/ReactIs.js]: > Yes, they probably don't make a difference in a large scheme of things, but if we can make it more efficient, why not do it?

I don't think the specific comment this is in response to is really more efficient one way or another, in terms of runtime efficiency. And the complexity/code-size difference is minimal. This is why I said it felt like a nit.

Comment (gaearon) [reply packages/react-is/src/ReactIs.js]: See my response below. Again I agree it probably doesn't matter in practice. But since it's a core utility and it's easy for us to minimize property access I don't see a reason not to.

Comment (bvaughn) [reply packages/react-is/src/ReactIs.js]: Sure, the comment we're responding to is the implementation of `isAsyncMode`, so the runtime path of `typeOf` didn't seem relevant since it wasn't being used (just the helper getter functions).

It's fine. I'm about to push the change you suggested.

**Comment by bvaughn** [comment]
I don't really agree that the previous implementation was inelegant, but your point about more efficient property comparison seems reasonable so I made the change. Testing locally now, will push in a moment. 👍

**Comment by bvaughn** [comment]
Thanks for the reviews, Dan and Sub!

**Comment by ljharb** [comment]
@bvaughn `isElement` doesn't seem compatible with React 0.13 (https://github.com/airbnb/enzyme/issues/1538#issuecomment-400345286) - any chance that's an easy fix?

**Comment by gaearon** [comment]
Seems expected to me, https://github.com/facebook/react/pull/4832 didn't land until 0.14 so we didn't have `$$typeof`. Don't think we care about supporting 0.13 at this point.

**Comment by bvaughn** [comment]
Ah, thanks Dan!

I'm not really familiar with React 14 or 15 😅

**Comment by gaearon** [comment]
Lucky you!

**Comment by ljharb** [comment]
@gaearon in 0.13, isValidElement was https://github.com/facebook/react/blob/v0.13.3/src/classic/element/ReactElement.js#L290-L302 - if i made the tiny PR required to fall back to checking `_isReactElement` when `$$typeof` wasn't around, considering that fallback wouldn't require further maintenance, is there any chance about supporting it?

My hope is to migrate the final enzyme adapter using `React.isValidElement` to using `react-is`, if possible.

**Comment by gaearon** [comment]
It's an extra property lookup for a thing that isn't there, so I don't think we'll want to do this.

**Comment by ljharb** [comment]
What if I came up with a way to determine if $$typeof was present or not at require time, so there'd be no runtime cost?

**Comment by sebmarkbage** [comment]
I don't know if we've articulated whether this package has any versioning guarantees. In particular I'd like to preserve the option to use an `instanceof` check or other brand check in the future without doubling the size of this package.

**Comment by ljharb** [comment]
@sebmarkbage currently it works on 0.14, 15, and 16, fwiw. As far as extensibility - that's exactly why i want to use `react-is` instead of `React.isValidElement` everywhere - so that a package (like `airbnb-prop-types`) doesn't have to peerDepend on React in order to reliably detect elements. 0.13 support is the final piece (after some of the React.Children methods) I need to remove the React peerDep requirement from that package. Additionally, it'd be great to eliminate the one remaining enzyme adapter that calls `React.isValidElement`, just on a philosophical level.

I don't anticipate "0.13 support on isElement" to add more than a line or two, fwiw.
